### PR TITLE
Ben/chunking variations

### PR DESCRIPTION
**Pull request by benleetownsend**

### CODE REVIEW

File: finetune/input_pipeline.py [hunk 1]
```diff
@@ -411,6 +453,7 @@ def get_train_input_fns(self, Xs, Y=None, batch_size=None, val_size=None, contex
         train_dataset = (
             lambda: train_dataset_unbatched()
             .padded_batch(batch_size, padded_shapes=shapes, drop_remainder=False)
+            .cache()
```
Comment (benleetownsend) [inline finetune/input_pipeline.py]: remove

File: finetune/input_pipeline.py [hunk 2]
```diff
@@ -24,6 +24,44 @@
...
+        self.chunk_size = self.max_length - 2
+        self.useful_chunk_width = self.chunk_size - total_context_width 
+        self.justify = justify.lower()
+        
+        if self.justify == "l":
+            self.normal_start = 0
+        elif self.justify == "r":
+            self.normal_start = total_context_width
+        elif self.justify == "c":
+            self.normal_start =	total_context_width // 2
```
Comment (benleetownsend) [inline finetune/input_pipeline.py]: rogue tab instead of space????

Comment (madisonmay) [reply finetune/input_pipeline.py]: ghosts in your keyboard

File: finetune/config.py [hunk 3]
```diff
@@ -328,6 +328,8 @@ def get_default_config():
         multi_label_sequences=False,
         multi_label_threshold=0.5,
         chunk_long_sequences=True,
+        chunk_context=None,
+        chunk_alignment="c",
```
Comment (madisonmay) [inline finetune/config.py]: "left", "right", "center"

**Comment by madisonmay** [comment]
Other than the terminology for chunk_aligment this is good to merge IMO.

File: finetune/base.py [hunk 4]
```diff
@@ -74,6 +74,56 @@ def cleanup():
...
+        if config.optimize_for.lower() == "speed":
+            overrides = {
+                "max_length": 128 if config.chunk_long_sequences else config.base_model.settings["max_length"],
+                "n_epochs": 5,
+                "batch_size": 4,
+                "chunk_context": 16,
+            }
+            
+        elif config.optimize_for.lower() == "accuracy":
+            overrides =	{
```
Comment (madisonmay) [inline finetune/base.py]: rogue tab

File: finetune/base.py [hunk 5]
```diff
@@ -74,6 +74,56 @@ def cleanup():
             os.environ["TF_CPP_MIN_LOG_LEVEL"] = "0"
             tf_logging.set_verbosity(tf_logging.DEBUG)
 
+    def resolve_config(self, **kwargs):
+        assert_valid_config(**kwargs)
+        config = get_default_config()
+        config.base_model = kwargs.get("base_model", config.base_model)
+        auto_keys = []
+        for k, v in config.items():
+            if not isinstance(v, str) or v.lower() != "auto":
```
Comment (madisonmay) [inline finetune/base.py]: If v isn't a string it ends up in auto keys?

Comment (madisonmay) [reply finetune/base.py]: Wouldn't this break if someone set lm_coef or some other numeric value?

File: finetune/config.py [hunk 6]
```diff
@@ -245,19 +246,20 @@ def get_default_config():
...
-        max_length=512,
+        max_length="auto",
         weight_stddev=0.02,
         save_dtype=None,
         val_set=None,
         per_process_gpu_memory_fraction=None,
         adapter_size=None,  # from Parameter Efficient Transfer Learning paper
         distribution_strategy="central_storage",
         xla=False,
+        optimize_for="accuracy",
```
Comment (madisonmay) [inline finetune/config.py:265]: Would be nice to add a doc string for this one.

File: finetune/datasets/reuters.py [hunk 7]
```diff
@@ -78,8 +79,9 @@ def download(self):
         test_size=0.2,
         random_state=42
     )
-    model = SequenceLabeler(batch_size=1, n_epochs=3, val_size=0.0, max_length=512, chunk_long_sequences=True, subtoken_predictions=False, crf_sequence_labeling=True, multi_label_sequences=False)
+    model = SequenceLabeler(base_model=RoBERTa, optimize_for="speed")#, batch_size=4, n_epochs=8, val_size=0.0, max_length=128, chunk_long_sequences=True, subtoken_predictions=False, crf_sequence_labeling=True, multi_label_sequences=False, chunk_context=16)
```
Comment (madisonmay) [inline finetune/datasets/reuters.py]: Remove file from commit

File: speed_benchmarks/run.py [hunk 8]
```diff
@@ -0,0 +1,51 @@
+import time
+
+from tabulate import tabulate
```
Comment (madisonmay) [inline speed_benchmarks/run.py:3]: I don't think this new dependency is declared.

File: speed_benchmarks/run.py [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+    x, y = sequence_data()
+    return benchmark(SequenceLabeler, config, x, y, runs=runs)
+
+def benchmark_multi_sequence(config, runs=5):
+    x, y = multi_label_sequence_data()
+    config["multi_label_sequences"] = True
+    return benchmark(SequenceLabeler, config, x, y, runs=runs)
+
+if __name__ == "__main__":
+    runs = 1
```
Comment (madisonmay) [inline speed_benchmarks/run.py]: Beautiful

File: requirements.txt [hunk 10]
```diff
@@ -15,3 +15,4 @@ regex>=2019.03.12
 lxml>=4.3.3
 tqdl==0.0.4
 sentencepiece>=0.1.83
+tabulate>=0.8.6
```
Comment (madisonmay) [inline requirements.txt]: Would set to tabulate>=0.8.6,<0.9.0 in both locations.

**Comment by benleetownsend** [comment]
@madisonmay This is ready to merge as far as i'm concerned.

**Pull request merged**